File: cross-project-tests/debuginfo-tests/dexter/dex/utils/Logging.py
# DExTer : Debugging Experience Tester
# ~~~~~~   ~         ~~         ~   ~~
#
# Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
# See https://llvm.org/LICENSE.txt for license information.
# SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
"""Utility functions for producing command line warnings."""

from dex.utils import PrettyOutput


class Logger(object):
    def __init__(self, pretty_output: PrettyOutput):
        self.o = pretty_output
        self.error_color = self.o.red
        self.warning_color = self.o.yellow
        self.note_color = self.o.default
        self.verbosity = 1

    def error(self, msg, enable_prefix=True, flag=None):
        if self.verbosity < 0:
            return
        if enable_prefix:
            msg = f"error: {msg}"
        if flag:
            msg = f"{msg} <y>[{flag}]</>"
        self.error_color("{}\n".format(msg), stream=PrettyOutput.stderr)

    def warning(self, msg, enable_prefix=True, flag=None):
        if self.verbosity < 1:
            return
        if enable_prefix:
            msg = f"warning: {msg}"
        if flag:
            msg = f"{msg} <y>[{flag}]</>"
        self.warning_color("{}\n".format(msg), stream=PrettyOutput.stderr)

    def note(self, msg, enable_prefix=True, flag=None):
        if self.verbosity < 2:
            return
        if enable_prefix:
            msg = f"note: {msg}"
        if flag:
            msg = f"{msg} <y>[{flag}]</>"
        self.note_color("{}\n".format(msg), stream=PrettyOutput.stderr)

File: llvm/tools/llvm-advisor/src/Utils/Logging.cpp
//===------------------- Logging.cpp - LLVM Advisor -------------------===//
//
// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
// See https://llvm.org/LICENSE.txt for license information.
// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
//
//===----------------------------------------------------------------------===//
//
// Structured logging with multiple severity levels.
// Provides logging infrastructure for the entire advisor.
//
//===----------------------------------------------------------------------===//
#include "Utils/Logging.h"

using namespace llvm;
using namespace llvm::advisor;

static StringRef levelLabel(LogLevel Level) {
  switch (Level) {
  case LogLevel::Error:
    return "error";
  case LogLevel::Warning:
    return "warning";
  case LogLevel::Info:
    return "info";
  case LogLevel::Debug:
    return "debug";
  }
  llvm_unreachable("Unknown log level");
}

void Logger::log(LogLevel Level, StringRef Message) {
  if (Level == LogLevel::Debug && !Verbose)
    return;
  OS << "llvm-advisor: " << levelLabel(Level) << ": " << Message << '\n';
}
